# Step 18 — ADP-A v2 Dataset Preparation
**Phase:** 4 — Model Training & Adapter Alignment  
**Spec authority:** SPEC-400 §3.2, §7.0.1, §7.0.3; REQ-400-LF2, REQ-400-CL1, REQ-400-171  
**Prerequisites:** Step 17 complete (`finetuning/adp_c_evaluator/adp_c_v2_final/` exists and smoke-tested)  
**Input:** Five open-licence corpora (same sources as Step 13)  
**Output:** `finetuning/adp_a_empathy/data/adp_a_v2_train.jsonl`  
**Supersedes:** Step 13 (`adp_a_train.jsonl` — 646 records, preserved for rollback)

---

## Why this notebook exists

Step 13 produced 646 training records for ADP-A. Post-hoc audit (Phase 4, 2026-05-13) identified three quality problems in that corpus:

| Problem | Count | Impact |
|---------|-------|--------|
| Fragment inputs (<10 words) — decontextualised therapy transcript turns | 124 | Model receives no learnable context |
| Short outputs (<15 words) — closers, paraphrases, generic acknowledgements | 90 | Insufficient empathy signal |
| URL-containing outputs | 14 | ADP-C v2 flags these FAIL at evaluation time |

This notebook rebuilds the corpus with:
1. **Enhanced `is_clean()` filter** — catches all three problem classes before records enter the training set.
2. **Expanded candidate caps and targets** — higher extraction caps offset the tighter filter, targeting ~1 300 final records.
3. **ADP-C v2 as the rejection-sampling oracle** — uses the retrained evaluator (Step 17) which correctly handles organic empathy data (G-DATA-06 fix).

## Dataset Mix

Weights are unchanged from REQ-400-LF2. Only candidate caps and targets are scaled up.

| Dataset | Weight | Candidate cap | Target | Step 13 target |
|---------|--------|--------------|--------|----------------|
| AnnoMI | 30% | 1 200 | 400 | 240 |
| Amod | 25% | 1 000 | 320 | 200 |
| ESConv | 20% | 800 | 250 | 160 |
| MentalChat16K | 15% | 600 | 200 | 120 |
| EmpatheticDialogues | 10% | 400 | 130 | 80 |
| **Total** | | | **~1 300** | **800** |

## ADP-C v2 Rejection Sampling

ADP-C v2 (Step 17) is used as the quality oracle. `ACCEPT_THRESHOLD` remains 0.50 (Director-ratified 2026-05-11, G-DATA-06). AnnoMI retains its ADP-C exemption — calibration mismatch between MI dialogue style and ADP-C training distribution is unresolved at this threshold.

## Sections
0. Environment + imports  
1. Configuration  
2. Load ADP-C v2 evaluator  
3. Quality filters  
4. Dataset loaders  
5. ADP-C rejection sampling  
6. Assembly & AnnoMI top-up  
7. Dataset audit & validation

## Section 0 — Environment + Imports

In [1]:
import os

# CRITICAL: set before any CUDA op — prevents WDDM memory fragmentation OOM on Windows.
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"
os.environ["TOKENIZERS_PARALLELISM"]   = "false"

# CRITICAL import order on Windows: datasets and trl BEFORE torch.
# Both register pyarrow workers at import time; if torch initialises CUDA first,
# pyarrow's multiprocessing fork fails with a CUDA re-init error.
import datasets
import trl
import torch

print(f"CUDA available : {torch.cuda.is_available()}")
print(f"Device         : {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'CPU'}")

C:\Users\dalet\anaconda3\envs\nikko\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


CUDA available : True
Device         : NVIDIA GeForce RTX 3070


In [2]:
import json
import random
import re
import unicodedata
import pathlib
from datasets import load_dataset
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
from peft import PeftModel

SEED = 42
random.seed(SEED)
torch.manual_seed(SEED)

## Section 1 — Configuration

In [3]:
# ── Paths ─────────────────────────────────────────────────────────────────────
BASE_DIR   = pathlib.Path(r"D:/Git Repos/nikko-companion")
FINETUNING = BASE_DIR / "finetuning"

# Step 17 output — ADP-C v2 is the rejection sampling oracle for this run.
# ADP-C v1 (adp_c_final/) is preserved for rollback but NOT used here.
ADP_C_DIR  = FINETUNING / "adp_c_evaluator" / "adp_c_v2_final"

# Step 13 output is preserved at adp_a_train.jsonl — do not overwrite it.
OUT_DIR    = FINETUNING / "adp_a_empathy" / "data"
OUT_FILE   = OUT_DIR / "adp_a_v2_train.jsonl"
OUT_DIR.mkdir(parents=True, exist_ok=True)

print(f"ADP-C v2 adapter : {ADP_C_DIR}")
print(f"Output file      : {OUT_FILE}")
assert ADP_C_DIR.exists(), f"ADP-C v2 adapter not found at {ADP_C_DIR} — run Step 17 first."

# ── Rejection sampling threshold ──────────────────────────────────────────────
# 0.50 — Director-ratified 2026-05-11 (G-DATA-06). Organic empathy datasets
# scored 1-12% at 0.70; 0.50 retains meaningful filtering without penalising
# natural conversational register.
ACCEPT_THRESHOLD = 0.50

# ── Dataset mix — weights from REQ-400-LF2, caps/targets scaled up from Step 13 ──
DATASET_MIX = [
    # Caps raised 2026-05-14: generous buffers added above target to account for
    # ADP-C pass rates running ~50% below expected in practice (G-DATA-06 audit).
    {"id": "annomi",     "weight": 0.30, "candidate_cap": 1500, "target": 220,
     "hf_name": "to-be/annomi-motivational-interviewing-therapy-conversations"},
    {"id": "amod",       "weight": 0.25, "candidate_cap": 3000, "target": 600,
     "hf_name": "Amod/mental_health_counseling_conversations"},
    {"id": "esconv",     "weight": 0.20, "candidate_cap": 3000, "target": 180,
     "hf_name": "thu-coai/esconv"},
    {"id": "mentalchat", "weight": 0.15, "candidate_cap": 2500, "target": 750,
     "hf_name": "ShenLab/MentalChat16K"},
    {"id": "empathetic", "weight": 0.10, "candidate_cap": 1500, "target": 120,
     "hf_name": "facebook/empathetic_dialogues"},
]

total_target = sum(d["target"] for d in DATASET_MIX)
print(f"Total target records : {total_target}  (Step 13 was 800)")
print("Configuration loaded.")

ADP-C v2 adapter : D:\Git Repos\nikko-companion\finetuning\adp_c_evaluator\adp_c_v2_final
Output file      : D:\Git Repos\nikko-companion\finetuning\adp_a_empathy\data\adp_a_v2_train.jsonl
Total target records : 1870  (Step 13 was 800)
Configuration loaded.


## Section 2 — Load ADP-C v2 Evaluator

ADP-C v2 (Step 17) is loaded as the rejection-sampling oracle. It is loaded once here
and reused across all five dataset filter passes in Section 5. The base model is 4-bit
NF4 quantised to fit within the RTX 3070 VRAM budget; only the adapter weights are
kept in bfloat16 on top.

Key difference from Step 13: this run uses `adp_c_v2_final/` which was retrained in
Step 17 with MULTI-TURN and URL-HALLUC redlines and organic PASS calibration examples.
It scores organic empathy data significantly more accurately than v1.

In [4]:
BASE_MODEL_ID = "mistralai/Mistral-7B-Instruct-v0.3"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_use_double_quant=True,
)

print("Loading tokenizer...")
tokenizer = AutoTokenizer.from_pretrained(BASE_MODEL_ID)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

print("Loading base model (4-bit NF4)...")
base_model = AutoModelForCausalLM.from_pretrained(
    BASE_MODEL_ID,
    quantization_config=bnb_config,
    device_map="auto",
    max_memory={0: "5000MiB", "cpu": "16GiB"},
    torch_dtype=torch.bfloat16,
)

print(f"Loading ADP-C v2 adapter from {ADP_C_DIR}...")
adp_c_model = PeftModel.from_pretrained(base_model, str(ADP_C_DIR))
adp_c_model.eval()
print("ADP-C v2 ready.")
print(f"VRAM used: {torch.cuda.memory_allocated(0) / 1024**3:.2f} GB")

Loading tokenizer...


Loading base model (4-bit NF4)...


Loading checkpoint shards:   0%|                                                                 | 0/3 [00:00<?, ?it/s]

C:\Users\dalet\anaconda3\envs\nikko\Lib\site-packages\accelerate\utils\modeling.py:329: UserWarning: expandable_segments not supported on this platform (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\c10/cuda/CUDAAllocatorConfig.h:28.)
  new_value = value.to(device)


Loading checkpoint shards:  33%|███████████████████                                      | 1/3 [00:05<00:11,  5.70s/it]

Loading checkpoint shards:  67%|██████████████████████████████████████                   | 2/3 [00:11<00:05,  5.72s/it]

Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████| 3/3 [00:16<00:00,  5.42s/it]

Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████| 3/3 [00:16<00:00,  5.50s/it]

Loading ADP-C v2 adapter from D:\Git Repos\nikko-companion\finetuning\adp_c_evaluator\adp_c_v2_final...


ADP-C v2 ready.
VRAM used: 3.91 GB


In [5]:
def adp_c_score(instruction: str, response: str) -> float:
    """
    Score a (instruction, response) pair using ADP-C v2.
    Returns float in [0, 1]. Records scoring >= ACCEPT_THRESHOLD are kept.

    The prompt asks ADP-C to rate empathy quality on a 0-1 scale.
    ADP-C v2 was calibrated on organic empathy data (G-DATA-06) so its
    PASS rate on natural conversational register is significantly higher
    than v1 (~6% on AnnoMI → expected ~40-60% post-calibration).
    """
    prompt = (
        f"[INST] Rate the empathy quality of this response on a scale of 0 to 1.\n"
        f"Instruction: {instruction}\n"
        f"Response: {response}\n"
        f"Reply with only a decimal number between 0 and 1. [/INST]"
    )
    inputs = tokenizer(prompt, return_tensors="pt", truncation=True, max_length=512).to("cuda")
    with torch.no_grad():
        output = adp_c_model.generate(
            **inputs,
            max_new_tokens=8,
            do_sample=False,
            pad_token_id=tokenizer.eos_token_id,
        )
    generated = tokenizer.decode(
        output[0][inputs["input_ids"].shape[1]:],
        skip_special_tokens=True,
    ).strip()
    match = re.search(r"[0-9]+\.?[0-9]*", generated)
    if match:
        return min(max(float(match.group()), 0.0), 1.0)
    return 0.0

## Section 3 — Quality Filters

The `is_clean()` function runs after ADP-C scoring and catches artifact classes
that the scorer passes but that degrade training quality. Three new rules added
versus Step 13:

| Filter | Step 13 threshold | Step 18 threshold | Rationale |
|--------|-------------------|-------------------|----------|
| Min instruction length | 8 words | 10 words | Eliminates decontextualised fragment turns (e.g. "Yeah.", "on facebook message as well") |
| Min output length | 8 words | 15 words | Eliminates closers and paraphrases that carry no empathy signal |
| URL filter | absent | blocks http/www/domain patterns | Prevents URL-HALLUC failures at ADP-C evaluation time |

In [6]:
# ── Turn-marker pattern — unchanged from Step 13 ──────────────────────────────
TURN_MARKERS = re.compile(
    r"(^|\s)(User:|Nikko:|Human:|Assistant:|Client:|Therapist:)︙|"
    r"\[Operating Mode|<s>|\[INST\]|\[/INST\]",
    re.IGNORECASE,
)

# ── URL / email pattern — NEW in Step 18 ──────────────────────────────────────
# Catches: https://, http://, www., bare domains (domain.com / .org / .au / .gov / .net),
# and email addresses (@). Phone numbers (13 11 14, 000) are not matched.
URL_PATTERN = re.compile(
    r"https?://|www\.\S+|@[a-zA-Z]|\b\w+\.(com|org|au|gov|net|edu)\b",
    re.IGNORECASE,
)

def is_english(text: str) -> bool:
    """Reject records where >15% of letter characters are non-Latin."""
    if not text:
        return False
    non_latin = sum(
        1 for c in text
        if unicodedata.category(c).startswith("L") and ord(c) > 591
    )
    return (non_latin / max(len(text), 1)) < 0.15


def is_clean(record: dict) -> bool:
    """
    Post-ADP-C quality gate. Returns True only if the record passes all filters.
    Called on every candidate before it enters the training set assembly.

    Filter order (fail-fast — cheapest checks first):
      1. Min word counts (instruction >= 10, output >= 15)
      2. URL / email in output
      3. Turn markers in either field
      4. Non-English characters
    """
    instruction = record.get("instruction", "")
    output      = record.get("output", "")

    # 1. Minimum word counts
    # 10-word floor on instruction eliminates decontextualised therapy fragments
    # (e.g. single-word back-channels like "Yeah.", "Okay.").
    # 15-word floor on output ensures enough empathy signal for the model to
    # learn from — closers and paraphrases under this threshold added noise in v1.
    if len(instruction.split()) < 10:
        return False
    if len(output.split()) < 15:
        return False

    # 2. URL / email in output — ADP-C v2 flags these FAIL at evaluation time.
    # Filtering here prevents contaminated examples from entering training.
    if URL_PATTERN.search(output):
        return False

    # 3. Multi-turn markers — same rule as Step 13.
    if TURN_MARKERS.search(output) or TURN_MARKERS.search(instruction):
        return False

    # 4. Non-English text
    if not is_english(output) or not is_english(instruction):
        return False

    return True


print("Quality filter functions defined.")
print("  is_clean() rules: min_instruction=10w, min_output=15w, no URLs, no turn markers, English only.")

Quality filter functions defined.
  is_clean() rules: min_instruction=10w, min_output=15w, no URLs, no turn markers, English only.


## Section 4 — Dataset Loaders

All five loaders are carried forward from Step 13 with schema fixes intact.
Candidate caps are increased — see Section 1 config. No loader logic changes.

In [7]:
# ── AnnoMI ─────────────────────────────────────────────────────────────────────
# Expert-annotated motivational interviewing transcripts.
# Schema: ShareGPT-style {'from': role, 'value': text} per turn.
# Short-form 'AnnoMI' returns 404 — must use full dataset ID (G-DATA-03).

CLIENT_ROLES    = {"client", "human", "user", "usr", "patient"}
THERAPIST_ROLES = {"therapist", "assistant", "gpt", "counselor", "counsellor", "sys"}

def load_annomi(cfg):
    ds = load_dataset(cfg["hf_name"], split="train", trust_remote_code=True)
    pairs = []
    for row in ds:
        convs = row.get("conversations") or []
        for i in range(len(convs) - 1):
            prev, curr = convs[i], convs[i + 1]
            if (prev.get("from") or "").lower().strip() not in CLIENT_ROLES:
                continue
            if (curr.get("from") or "").lower().strip() not in THERAPIST_ROLES:
                continue
            user_msg  = (prev.get("value") or prev.get("text") or "").strip()
            therapist = (curr.get("value") or curr.get("text") or "").strip()
            if not user_msg or not therapist:
                continue
            pairs.append({"instruction": user_msg, "output": therapist})
            if len(pairs) >= cfg["candidate_cap"]:
                break
        if len(pairs) >= cfg["candidate_cap"]:
            break
    random.shuffle(pairs)
    print(f"[annomi] {len(pairs)} candidate pairs extracted (cap={cfg['candidate_cap']})")
    return pairs

annomi_candidates = load_annomi(DATASET_MIX[0])

[annomi] 1500 candidate pairs extracted (cap=1500)


In [8]:
# ── Amod / mental_health_counseling_conversations ─────────────────────────────
# Real anonymised Q&A from licensed counselors. RAIL-D licence.
# Schema fix: LICENSE-RAIL-D.txt causes DatasetGenerationCastError when the
# builder picks it up alongside data files — glob to *.json only.

def load_amod(cfg):
    ds = None
    for kwargs in [
        {"data_files": {"train": "*.json"}, "split": "train"},
        {"data_files": {"train": "combined_dataset.json"}, "split": "train"},
        {"data_files": {"train": "data.json"}, "split": "train"},
        {"data_files": {"train": "train.json"}, "split": "train"},
    ]:
        try:
            ds = load_dataset(cfg["hf_name"], trust_remote_code=True, **kwargs)
            print(f"[amod] Loaded via {kwargs}")
            break
        except Exception as e:
            print(f"[amod] Strategy failed: {e}")
    if ds is None:
        print("[amod] All strategies failed — skipping")
        return []
    pairs = []
    for row in ds:
        instruction = (row.get("Context") or row.get("context") or row.get("question") or "").strip()
        output      = (row.get("Response") or row.get("response") or row.get("answer") or "").strip()
        if not instruction or not output:
            continue
        pairs.append({"instruction": instruction, "output": output})
        if len(pairs) >= cfg["candidate_cap"]:
            break
    random.shuffle(pairs)
    print(f"[amod] {len(pairs)} candidate pairs extracted (cap={cfg['candidate_cap']})")
    return pairs

amod_candidates = load_amod(DATASET_MIX[1])

[amod] Loaded via {'data_files': {'train': '*.json'}, 'split': 'train'}
[amod] 3000 candidate pairs extracted (cap=3000)


In [9]:
# ── ESConv (Emotional Support Conversation) ───────────────────────────────────
# ACL 2021 dataset; strategy-labelled utterances teach support sequencing.
# Schema fix: HF version wraps each conversation as a JSON string in a 'text'
# column rather than exposing a flat dict — parse text -> dict -> dialog list.

def load_esconv(cfg):
    ds = load_dataset(cfg["hf_name"], split="train", trust_remote_code=True)
    pairs = []
    for conv in ds:
        raw = conv.get("dialog") or conv.get("text") or ""
        if isinstance(raw, str):
            try:
                parsed = json.loads(raw)
            except json.JSONDecodeError:
                continue
            dialog = parsed.get("dialog", []) if isinstance(parsed, dict) else (
                parsed if isinstance(parsed, list) else []
            )
        elif isinstance(raw, list):
            dialog = raw
        else:
            continue
        for i in range(len(dialog) - 1):
            turn_a, turn_b = dialog[i], dialog[i + 1]
            spk_a = (turn_a.get("speaker") or turn_a.get("role") or "").lower()
            spk_b = (turn_b.get("speaker") or turn_b.get("role") or "").lower()
            if spk_a not in ("usr", "user", "human") or spk_b not in ("sys", "system", "assistant"):
                continue
            user_msg = (turn_a.get("text") or turn_a.get("content") or "").strip()
            support  = (turn_b.get("text") or turn_b.get("content") or "").strip()
            if not user_msg or not support:
                continue
            pairs.append({"instruction": user_msg, "output": support})
            if len(pairs) >= cfg["candidate_cap"]:
                break
        if len(pairs) >= cfg["candidate_cap"]:
            break
    random.shuffle(pairs)
    print(f"[esconv] {len(pairs)} candidate pairs extracted (cap={cfg['candidate_cap']})")
    return pairs

esconv_candidates = load_esconv(DATASET_MIX[2])

[esconv] 3000 candidate pairs extracted (cap=3000)


In [10]:
# ── MentalChat16K — synthetic multi-turn counselor/client conversations ────────
# Cleanest schema in the stack. REQ-400-171: must pass ADP-C filter.

def load_mentalchat(cfg):
    ds = load_dataset(cfg["hf_name"], split="train", trust_remote_code=True)
    pairs = []
    for row in ds:
        instruction = (row.get("input") or row.get("instruction") or row.get("question") or "").strip()
        output      = (row.get("output") or row.get("response") or row.get("answer") or "").strip()
        if not instruction or not output:
            continue
        pairs.append({"instruction": instruction, "output": output})
        if len(pairs) >= cfg["candidate_cap"]:
            break
    random.shuffle(pairs)
    print(f"[mentalchat] {len(pairs)} candidate pairs extracted (cap={cfg['candidate_cap']})")
    return pairs

mentalchat_candidates = load_mentalchat(DATASET_MIX[3])

[mentalchat] 2500 candidate pairs extracted (cap=2500)


In [11]:
# ── EmpatheticDialogues (Facebook AI) ────────────────────────────────────────
# Schema fix: utterance_idx is non-zero in current HF version — index filter
# would skip all rows. Replaced with conv_id deduplication (first row per
# conversation) which is stable across dataset version bumps.

def load_empathetic(cfg):
    ds = load_dataset(cfg["hf_name"], split="train", trust_remote_code=True)
    seen_convs = set()
    pairs = []
    for row in ds:
        conv_id = row.get("conv_id", "")
        if conv_id in seen_convs:
            continue
        seen_convs.add(conv_id)
        user_msg = (row.get("prompt") or "").strip().replace("_comma_", ",")
        response = (row.get("utterance") or "").strip().replace("_comma_", ",")
        if not user_msg or not response:
            continue
        pairs.append({"instruction": user_msg, "output": response})
        if len(pairs) >= cfg["candidate_cap"]:
            break
    random.shuffle(pairs)
    print(f"[empathetic] {len(pairs)} candidate pairs extracted (cap={cfg['candidate_cap']})")
    return pairs

empathetic_candidates = load_empathetic(DATASET_MIX[4])

[empathetic] 1500 candidate pairs extracted (cap=1500)


## Section 5 — ADP-C v2 Rejection Sampling

Each candidate pool is passed through `adp_c_score()`. Pairs scoring below
`ACCEPT_THRESHOLD` (0.50) are discarded. `is_clean()` is also run at this stage
to avoid scoring records that will be rejected anyway.

AnnoMI is run through the scorer for pass-rate logging only — its filtered
results are **not used** in assembly. AnnoMI is top-filled directly to target
in Section 6 (G-DATA-06 exemption).

In [12]:
all_candidates = [
    (DATASET_MIX[0], annomi_candidates),
    (DATASET_MIX[1], amod_candidates),
    (DATASET_MIX[2], esconv_candidates),
    (DATASET_MIX[3], mentalchat_candidates),
    (DATASET_MIX[4], empathetic_candidates),
]

filtered = {}

for cfg, candidates in all_candidates:
    ds_id    = cfg["id"]
    accepted = []
    skipped  = 0   # records failing is_clean() before scoring

    for i, pair in enumerate(candidates):
        # Run quality filter first — avoids scoring records we will reject anyway.
        if not is_clean(pair):
            skipped += 1
            continue
        score = adp_c_score(pair["instruction"], pair["output"])
        if score >= ACCEPT_THRESHOLD:
            accepted.append(pair)
        if (i + 1) % 50 == 0:
            pct = 100 * len(accepted) / max(i + 1 - skipped, 1)
            print(f"  [{ds_id}] {i+1}/{len(candidates)}  clean_pass={len(accepted)} ({pct:.0f}%)  skipped={skipped}...")

    total_scored = len(candidates) - skipped
    pass_pct     = 100 * len(accepted) / max(total_scored, 1)
    print(f"  [{ds_id}] Done. Scored={total_scored}  Pass={len(accepted)} ({pass_pct:.1f}%)  Pre-filter skip={skipped}")
    filtered[ds_id] = accepted

  [annomi] 50/1500  clean_pass=1 (10%)  skipped=40...


  [annomi] 200/1500  clean_pass=2 (6%)  skipped=165...


  [annomi] 1450/1500  clean_pass=41 (17%)  skipped=1212...


  [annomi] Done. Scored=253  Pass=42 (16.6%)  Pre-filter skip=1247


  [amod] 50/3000  clean_pass=20 (43%)  skipped=3...


  [amod] 100/3000  clean_pass=35 (36%)  skipped=3...


  [amod] 150/3000  clean_pass=46 (33%)  skipped=12...


  [amod] 250/3000  clean_pass=74 (32%)  skipped=19...


  [amod] 300/3000  clean_pass=90 (32%)  skipped=23...


  [amod] 350/3000  clean_pass=106 (33%)  skipped=32...


  [amod] 400/3000  clean_pass=118 (33%)  skipped=39...


  [amod] 450/3000  clean_pass=133 (33%)  skipped=44...


  [amod] 500/3000  clean_pass=150 (33%)  skipped=48...


  [amod] 600/3000  clean_pass=171 (32%)  skipped=59...


  [amod] 650/3000  clean_pass=187 (32%)  skipped=63...


  [amod] 700/3000  clean_pass=198 (31%)  skipped=69...


  [amod] 750/3000  clean_pass=213 (32%)  skipped=74...


  [amod] 800/3000  clean_pass=222 (31%)  skipped=80...


  [amod] 850/3000  clean_pass=239 (31%)  skipped=86...


  [amod] 900/3000  clean_pass=250 (31%)  skipped=91...


  [amod] 1000/3000  clean_pass=283 (31%)  skipped=99...


  [amod] 1100/3000  clean_pass=323 (32%)  skipped=105...


  [amod] 1150/3000  clean_pass=337 (32%)  skipped=109...


  [amod] 1250/3000  clean_pass=366 (32%)  skipped=120...


  [amod] 1300/3000  clean_pass=382 (32%)  skipped=124...


  [amod] 1350/3000  clean_pass=399 (33%)  skipped=127...


  [amod] 1400/3000  clean_pass=414 (33%)  skipped=132...


  [amod] 1450/3000  clean_pass=426 (32%)  skipped=135...


  [amod] 1500/3000  clean_pass=433 (32%)  skipped=142...


  [amod] 1550/3000  clean_pass=442 (31%)  skipped=145...


  [amod] 1600/3000  clean_pass=451 (31%)  skipped=149...


  [amod] 1650/3000  clean_pass=465 (31%)  skipped=153...


  [amod] 1700/3000  clean_pass=480 (31%)  skipped=157...


  [amod] 1750/3000  clean_pass=494 (31%)  skipped=162...


  [amod] 1800/3000  clean_pass=512 (31%)  skipped=167...


  [amod] 1850/3000  clean_pass=522 (31%)  skipped=175...


  [amod] 1900/3000  clean_pass=542 (31%)  skipped=179...


  [amod] 1950/3000  clean_pass=552 (31%)  skipped=183...


  [amod] 2050/3000  clean_pass=577 (31%)  skipped=195...


  [amod] 2100/3000  clean_pass=588 (31%)  skipped=201...


  [amod] 2150/3000  clean_pass=608 (31%)  skipped=206...


  [amod] 2250/3000  clean_pass=642 (32%)  skipped=215...


  [amod] 2300/3000  clean_pass=664 (32%)  skipped=221...


  [amod] 2350/3000  clean_pass=678 (32%)  skipped=223...


  [amod] 2400/3000  clean_pass=691 (32%)  skipped=225...


  [amod] 2500/3000  clean_pass=717 (32%)  skipped=237...


  [amod] 2550/3000  clean_pass=740 (32%)  skipped=240...


  [amod] 2600/3000  clean_pass=754 (32%)  skipped=243...


  [amod] 2650/3000  clean_pass=771 (32%)  skipped=249...


  [amod] 2700/3000  clean_pass=786 (32%)  skipped=254...


  [amod] 2750/3000  clean_pass=797 (32%)  skipped=261...


  [amod] 2800/3000  clean_pass=812 (32%)  skipped=264...


  [amod] 2850/3000  clean_pass=825 (32%)  skipped=267...


  [amod] 2900/3000  clean_pass=839 (32%)  skipped=271...


  [amod] 3000/3000  clean_pass=867 (32%)  skipped=279...
  [amod] Done. Scored=2721  Pass=867 (31.9%)  Pre-filter skip=279


  [esconv] 150/3000  clean_pass=14 (27%)  skipped=99...


  [esconv] 250/3000  clean_pass=19 (21%)  skipped=159...


  [esconv] 450/3000  clean_pass=31 (20%)  skipped=292...


  [esconv] 550/3000  clean_pass=37 (19%)  skipped=352...


  [esconv] 650/3000  clean_pass=45 (19%)  skipped=409...


  [esconv] 750/3000  clean_pass=52 (19%)  skipped=472...


  [esconv] 850/3000  clean_pass=61 (19%)  skipped=531...


  [esconv] 900/3000  clean_pass=64 (19%)  skipped=560...


  [esconv] 950/3000  clean_pass=72 (20%)  skipped=588...


  [esconv] 1050/3000  clean_pass=77 (19%)  skipped=645...


  [esconv] 1150/3000  clean_pass=82 (18%)  skipped=703...


  [esconv] 1250/3000  clean_pass=88 (18%)  skipped=766...


  [esconv] 1300/3000  clean_pass=92 (18%)  skipped=799...


  [esconv] 1450/3000  clean_pass=104 (18%)  skipped=884...


  [esconv] 1500/3000  clean_pass=111 (19%)  skipped=912...


  [esconv] 1700/3000  clean_pass=126 (19%)  skipped=1034...


  [esconv] 1750/3000  clean_pass=129 (19%)  skipped=1063...


  [esconv] 1850/3000  clean_pass=132 (18%)  skipped=1129...


  [esconv] 1900/3000  clean_pass=134 (18%)  skipped=1162...


  [esconv] 1950/3000  clean_pass=139 (18%)  skipped=1194...


  [esconv] 2100/3000  clean_pass=148 (18%)  skipped=1282...


  [esconv] 2200/3000  clean_pass=158 (18%)  skipped=1336...


  [esconv] 2300/3000  clean_pass=165 (18%)  skipped=1399...


  [esconv] 2350/3000  clean_pass=168 (18%)  skipped=1432...


  [esconv] 2400/3000  clean_pass=171 (18%)  skipped=1463...


  [esconv] 2500/3000  clean_pass=183 (19%)  skipped=1524...


  [esconv] 2550/3000  clean_pass=186 (19%)  skipped=1555...


  [esconv] 2700/3000  clean_pass=197 (19%)  skipped=1646...


  [esconv] 2750/3000  clean_pass=201 (19%)  skipped=1668...


  [esconv] 3000/3000  clean_pass=215 (18%)  skipped=1817...
  [esconv] Done. Scored=1183  Pass=215 (18.2%)  Pre-filter skip=1817


  [mentalchat] 50/2500  clean_pass=37 (74%)  skipped=0...


  [mentalchat] 100/2500  clean_pass=74 (74%)  skipped=0...


  [mentalchat] 150/2500  clean_pass=112 (75%)  skipped=0...


  [mentalchat] 200/2500  clean_pass=138 (69%)  skipped=0...


  [mentalchat] 250/2500  clean_pass=177 (71%)  skipped=0...


  [mentalchat] 300/2500  clean_pass=217 (72%)  skipped=0...


  [mentalchat] 350/2500  clean_pass=255 (73%)  skipped=0...


  [mentalchat] 400/2500  clean_pass=293 (73%)  skipped=0...


  [mentalchat] 450/2500  clean_pass=329 (73%)  skipped=0...


  [mentalchat] 500/2500  clean_pass=361 (72%)  skipped=0...


  [mentalchat] 550/2500  clean_pass=391 (71%)  skipped=0...


  [mentalchat] 600/2500  clean_pass=431 (72%)  skipped=0...


  [mentalchat] 650/2500  clean_pass=468 (72%)  skipped=0...


  [mentalchat] 700/2500  clean_pass=504 (72%)  skipped=0...


  [mentalchat] 750/2500  clean_pass=540 (72%)  skipped=0...


  [mentalchat] 800/2500  clean_pass=574 (72%)  skipped=0...


  [mentalchat] 850/2500  clean_pass=610 (72%)  skipped=0...


  [mentalchat] 900/2500  clean_pass=644 (72%)  skipped=0...


  [mentalchat] 950/2500  clean_pass=682 (72%)  skipped=0...


  [mentalchat] 1000/2500  clean_pass=718 (72%)  skipped=0...


  [mentalchat] 1050/2500  clean_pass=750 (71%)  skipped=0...


  [mentalchat] 1100/2500  clean_pass=785 (71%)  skipped=0...


  [mentalchat] 1150/2500  clean_pass=817 (71%)  skipped=0...


  [mentalchat] 1200/2500  clean_pass=854 (71%)  skipped=0...


  [mentalchat] 1250/2500  clean_pass=892 (71%)  skipped=0...


  [mentalchat] 1300/2500  clean_pass=923 (71%)  skipped=0...


  [mentalchat] 1350/2500  clean_pass=963 (71%)  skipped=0...


  [mentalchat] 1400/2500  clean_pass=999 (71%)  skipped=0...


  [mentalchat] 1450/2500  clean_pass=1033 (71%)  skipped=0...


  [mentalchat] 1500/2500  clean_pass=1071 (71%)  skipped=0...


  [mentalchat] 1550/2500  clean_pass=1110 (72%)  skipped=0...


  [mentalchat] 1600/2500  clean_pass=1141 (71%)  skipped=0...


  [mentalchat] 1650/2500  clean_pass=1180 (72%)  skipped=0...


  [mentalchat] 1700/2500  clean_pass=1214 (71%)  skipped=0...


  [mentalchat] 1750/2500  clean_pass=1252 (72%)  skipped=0...


  [mentalchat] 1800/2500  clean_pass=1287 (72%)  skipped=0...


  [mentalchat] 1850/2500  clean_pass=1319 (71%)  skipped=0...


  [mentalchat] 1900/2500  clean_pass=1350 (71%)  skipped=0...


  [mentalchat] 1950/2500  clean_pass=1384 (71%)  skipped=0...


  [mentalchat] 2000/2500  clean_pass=1415 (71%)  skipped=0...


  [mentalchat] 2050/2500  clean_pass=1448 (71%)  skipped=0...


  [mentalchat] 2100/2500  clean_pass=1477 (70%)  skipped=0...


  [mentalchat] 2150/2500  clean_pass=1519 (71%)  skipped=0...


  [mentalchat] 2200/2500  clean_pass=1556 (71%)  skipped=0...


  [mentalchat] 2250/2500  clean_pass=1593 (71%)  skipped=0...


  [mentalchat] 2300/2500  clean_pass=1623 (71%)  skipped=1...


  [mentalchat] 2350/2500  clean_pass=1659 (71%)  skipped=1...


  [mentalchat] 2400/2500  clean_pass=1690 (70%)  skipped=1...


  [mentalchat] 2450/2500  clean_pass=1722 (70%)  skipped=1...


  [mentalchat] 2500/2500  clean_pass=1758 (70%)  skipped=1...
  [mentalchat] Done. Scored=2499  Pass=1758 (70.3%)  Pre-filter skip=1


  [empathetic] 50/1500  clean_pass=1 (4%)  skipped=25...


  [empathetic] 100/1500  clean_pass=8 (15%)  skipped=46...


  [empathetic] 200/1500  clean_pass=17 (16%)  skipped=94...


  [empathetic] 250/1500  clean_pass=19 (15%)  skipped=127...


  [empathetic] 400/1500  clean_pass=39 (19%)  skipped=194...


  [empathetic] 550/1500  clean_pass=54 (19%)  skipped=260...


  [empathetic] 600/1500  clean_pass=65 (20%)  skipped=281...


  [empathetic] 800/1500  clean_pass=84 (20%)  skipped=386...


  [empathetic] 950/1500  clean_pass=96 (20%)  skipped=466...


  [empathetic] 1000/1500  clean_pass=97 (19%)  skipped=485...


  [empathetic] 1100/1500  clean_pass=114 (20%)  skipped=534...


  [empathetic] 1150/1500  clean_pass=120 (20%)  skipped=557...


  [empathetic] 1300/1500  clean_pass=133 (20%)  skipped=640...


  [empathetic] 1350/1500  clean_pass=137 (20%)  skipped=664...


  [empathetic] 1400/1500  clean_pass=143 (20%)  skipped=688...


  [empathetic] 1500/1500  clean_pass=151 (20%)  skipped=740...
  [empathetic] Done. Scored=760  Pass=151 (19.9%)  Pre-filter skip=740


## Section 6 — Assembly & AnnoMI Top-up

Filtered pools are sampled down to each dataset's target and written to
`adp_a_v2_train.jsonl`. AnnoMI is then top-filled directly from the full
candidate pool (no ADP-C filter, `is_clean()` only) to reach its 400-record
target — G-DATA-06 exemption.

In [13]:
sampled_all = []
audit_rows  = []

for cfg in DATASET_MIX:
    ds_id     = cfg["id"]
    target    = cfg["target"]
    available = filtered.get(ds_id, [])
    sampled   = available[:target]
    shortfall = target - len(sampled)
    status    = f"SHORT by {shortfall}" if shortfall > 0 else "OK"

    # Tag each record with its source dataset for Step 19 distribution audit.
    for r in sampled:
        r["source"] = ds_id
    sampled_all.extend(sampled)
    audit_rows.append([ds_id, target, len(available), len(sampled), shortfall, status])

random.shuffle(sampled_all)

print(f"{'Dataset':<22} {'Target':>8} {'Filtered':>10} {'Sampled':>10} {'Shortfall':>10}  Status")
print("-" * 75)
for r in audit_rows:
    print(f"{r[0]:<22} {r[1]:>8} {r[2]:>10} {r[3]:>10} {r[4]:>10}  {r[5]}")
print("-" * 75)
print(f"{'PRE-TOPUP TOTAL':<22} {sum(c['target'] for c in DATASET_MIX):>8} {'':>10} {len(sampled_all):>10}")

# Write pre-topup records
with open(OUT_FILE, "w", encoding="utf-8") as f:
    for rec in sampled_all:
        f.write(json.dumps(rec, ensure_ascii=False) + "\n")
print(f"\nPre-topup records written: {len(sampled_all)}")

Dataset                  Target   Filtered    Sampled  Shortfall  Status
---------------------------------------------------------------------------
annomi                      220         42         42        178  SHORT by 178
amod                        600        867        600          0  OK
esconv                      180        215        180          0  OK
mentalchat                  750       1758        750          0  OK
empathetic                  120        151        120          0  OK
---------------------------------------------------------------------------
PRE-TOPUP TOTAL            1870                  1692

Pre-topup records written: 1692


In [14]:
# ── AnnoMI top-up — G-DATA-06 exemption ──────────────────────────────────────
# ADP-C pass rate for AnnoMI is low due to calibration mismatch between MI
# dialogue style and ADP-C's training distribution. Tier 1 expert-annotated
# source is filled directly to target using is_clean() only.

ANNOMI_TARGET  = DATASET_MIX[0]["target"]  # 400
annomi_sampled = next(r[3] for r in audit_rows if r[0] == "annomi")
topup_needed   = max(0, ANNOMI_TARGET - annomi_sampled)

print(f"AnnoMI in filtered set : {annomi_sampled}")
print(f"Target                 : {ANNOMI_TARGET}")
print(f"Top-up needed          : {topup_needed}")

# Guard: ensure file matches what we just wrote before appending.
existing_lines = [l for l in OUT_FILE.read_text(encoding="utf-8").splitlines() if l.strip()]
assert len(existing_lines) == len(sampled_all), (
    f"File has {len(existing_lines)} records but sampled_all has {len(sampled_all)} — "
    "re-run the assembly cell before re-running this cell."
)

if topup_needed > 0:
    already_texts = {r["instruction"] for r in filtered.get("annomi", [])}
    # Pool: AnnoMI candidates not already in filtered set, passing is_clean().
    topup_pool = [
        r for r in annomi_candidates
        if r["instruction"] not in already_texts and is_clean(r)
    ]
    random.shuffle(topup_pool)
    topup_records = topup_pool[:topup_needed]
    with open(OUT_FILE, "a", encoding="utf-8") as f:
        for rec in topup_records:
            rec["source"] = "annomi"
            f.write(json.dumps(rec, ensure_ascii=False) + "\n")
    print(f"Appended {len(topup_records)} AnnoMI top-up records (is_clean only, no ADP-C filter).")
else:
    print("AnnoMI already at target — no top-up needed.")

AnnoMI in filtered set : 42
Target                 : 220
Top-up needed          : 178
Appended 178 AnnoMI top-up records (is_clean only, no ADP-C filter).


## Section 7 — Dataset Audit & Validation

Final checks before handing off to Step 19. Hard assertions on missing fields;
soft warnings on shortfall. If any dataset is SHORT by more than 30%, consider
increasing its `candidate_cap` and rerunning from Section 4.

In [15]:
from collections import Counter

final_records = [json.loads(l) for l in OUT_FILE.read_text(encoding="utf-8").splitlines() if l.strip()]
file_size_kb  = OUT_FILE.stat().st_size / 1024

# Field completeness
missing_i = sum(1 for r in final_records if not r.get("instruction"))
missing_o = sum(1 for r in final_records if not r.get("output"))

# Source distribution
source_dist = Counter(r.get("source", "unknown") for r in final_records)

# Re-run is_clean() on final set — catches anything that slipped through
dirty = [i for i, r in enumerate(final_records) if not is_clean(r)]

# URL check on final set
url_hits = [i for i, r in enumerate(final_records) if URL_PATTERN.search(r.get("output", ""))]

print(f"Final total records  : {len(final_records)}")
print(f"File size            : {file_size_kb:.1f} KB")
print(f"Missing instruction  : {missing_i}")
print(f"Missing output       : {missing_o}")
print(f"Dirty records        : {len(dirty)}")
print(f"URL-containing output: {len(url_hits)}")
print()
print("Source distribution:")
for source, count in sorted(source_dist.items()):
    pct = 100 * count / len(final_records)
    target = next((d["target"] for d in DATASET_MIX if d["id"] == source), 0)
    print(f"  {source:<22} {count:>5}  ({pct:.1f}%)  target={target}")

print()
assert missing_i == 0, f"ABORT: {missing_i} records have empty instruction fields."
assert missing_o == 0, f"ABORT: {missing_o} records have empty output fields."
assert len(url_hits) == 0, f"ABORT: {len(url_hits)} records contain URLs in output. Inspect indices: {url_hits[:5]}"
assert len(dirty) == 0,    f"ABORT: {len(dirty)} records failed is_clean(). Inspect indices: {dirty[:5]}"

print("All validation gates passed.")
print(f"adp_a_v2_train.jsonl is ready for Step 19.")
print(f"Location: {OUT_FILE.resolve()}")

Final total records  : 1870
File size            : 2048.6 KB
Missing instruction  : 0
Missing output       : 0
Dirty records        : 0
URL-containing output: 0

Source distribution:
  amod                     600  (32.1%)  target=600
  annomi                   220  (11.8%)  target=220
  empathetic               120  (6.4%)  target=120
  esconv                   180  (9.6%)  target=180
  mentalchat               750  (40.1%)  target=750

All validation gates passed.
adp_a_v2_train.jsonl is ready for Step 19.
Location: D:\Git Repos\nikko-companion\finetuning\adp_a_empathy\data\adp_a_v2_train.jsonl
